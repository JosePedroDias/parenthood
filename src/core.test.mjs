import {
  handleWhitespaceAndComment,
  tokenizeByOperand,
  processPplString,
  pplStringToRelationships,
} from './core.mjs';

describe('handleWhitespaceAndComment', () => {
  it('basics', () => {
    expect(handleWhitespaceAndComment('')).toBe('');
    expect(handleWhitespaceAndComment('  ')).toBe('');
    expect(handleWhitespaceAndComment('\t\t')).toBe('');
    expect(handleWhitespaceAndComment('  a thing')).toBe('a thing');
    expect(handleWhitespaceAndComment('\t\tAnother one')).toBe('Another one');
    expect(handleWhitespaceAndComment('# just a comment')).toBe('');
    expect(handleWhitespaceAndComment('  a thing#stuffs')).toBe('a thing');
    expect(handleWhitespaceAndComment('\t\tAnother one# more stuffs')).toBe(
      'Another one'
    );
  });
});

describe('tokenizeByOperand', () => {
  it('basics', () => {
    expect(tokenizeByOperand('Adam+/Eve')).toEqual(['Adam', '+/', 'Eve']);
    expect(tokenizeByOperand('Ad am +/\tE ve')).toEqual([
      'Ad am',
      '+/',
      'E ve',
    ]);
    expect(tokenizeByOperand('Ad am +\tE ve')).toEqual(['Ad am', '+', 'E ve']);
    expect(tokenizeByOperand('Ad am ..\tE ve')).toEqual([
      'Ad am',
      '..',
      'E ve',
    ]);
    expect(tokenizeByOperand('Ad am @\tE ve')).toEqual(undefined);
  });
});

describe('pplStringToRelationships', () => {
  it('fails if no known operand is found', () => {
    expect(() => pplStringToRelationships(`p 1 @ p 2`)).toThrow(Error);
  });

  it('parses with relationship wo /siblings', () => {
    expect(pplStringToRelationships(`p 1 + p 2`)).toEqual([
      {
        a: 'p 1',
        b: 'p 2',
        kind: 'with',
      },
    ]);
  });

  it('parses almost_with relationship w/ 1 sibling', () => {
    expect(
      pplStringToRelationships(`p 1 .. p 2
  s 1`)
    ).toEqual([
      {
        a: 'p 1',
        b: 'p 2',
        kind: 'almost_with',
        siblings: ['s 1'],
      },
    ]);
  });

  it('parses has_been_with relationship w/ 2 siblings', () => {
    expect(
      pplStringToRelationships(`p1 +/ p2
\ts1
  s2`)
    ).toEqual([
      {
        a: 'p1',
        b: 'p2',
        kind: 'has_been_with',
        siblings: ['s1', 's2'],
      },
    ]);
  });

  it('parses almost_with relationship w/ 1 sibling and comments', () => {
    expect(
      pplStringToRelationships(`
#something
p1 .. p2
  # something else
  s1
# and even more`)
    ).toEqual([
      {
        a: 'p1',
        b: 'p2',
        kind: 'almost_with',
        siblings: ['s1'],
      },
    ]);
  });
});

describe('processPplString', () => {
  it('processes 1 has_been_with relationship w/ 2 siblings', () => {
    expect(
      processPplString(`p1 +/ p2
    s1
    s2`)
    ).toMatchSnapshot();
  });
});
