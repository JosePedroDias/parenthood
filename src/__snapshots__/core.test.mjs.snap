// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processPplString processes 1 has_been_with relationship w/ 2 siblings 1`] = `
"graph {
  node [shape=rect style=rounded fontname=Helvetica fontsize=10 margin=0.02]
  splines=true
  overlap=false
  #pack=true
  #packMode=\\"node\\"
  #ratio=1.5
  #rankdir=\\"LR\\"
  #newrank=true


  subgraph {
    rank=same
    \\"p1\\" -- \\"p1_p2\\" -- \\"p2\\" [ color=red penwidth=1.5 weight=2 len=0.7]
    \\"p1\\" [label=\\"p1\\" tooltip=\\"p1\\"]
    \\"p2\\" [label=\\"p2\\" tooltip=\\"p2\\"]
    \\"p1_p2\\" [ color=red shape=none width=0 height=0 label=\\"\\"]
  }

  \\"p1_p2\\" -- \\"p1_p2_sibs\\"

  \\"p1_p2_sibs\\" [shape=none width=0 height=0 label=\\"\\"]

  \\"p1_p2_sibs\\" -- \\"s1\\" [dir=forward arrowType=open arrowsize=0.5]
  \\"p1_p2_sibs\\" -- \\"s2\\" [dir=forward arrowType=open arrowsize=0.5]

  subgraph {
    rank=same
    \\"s1\\" [label=\\"s1\\" tooltip=\\"s1\\"]
    \\"s2\\" [label=\\"s2\\" tooltip=\\"s2\\"]
  }

}"
`;
